File: frontend/src/types/accounting.ts
/**
 * Accounting TypeScript Interfaces
 * Type definitions for accounting models matching backend structure
 */

// Enums
export type AccountType = 'asset' | 'liability' | 'equity' | 'revenue' | 'expense';
export type TransactionType = 'debit' | 'credit';
export type InvoiceStatus = 'draft' | 'sent' | 'paid' | 'overdue' | 'cancelled';

// Base interfaces
export interface BaseEntity {
  id: string;
  org_id: string;
  created_at: string;
  updated_at?: string;
}

// Account interface
export interface Account extends BaseEntity {
  account_number: string;
  account_name: string;
  account_type: AccountType;
  parent_account_id?: string;
  description?: string;
  is_active: boolean;
  balance: string; // Decimal as string - current account balance
}

// Transaction interface
export interface Transaction extends BaseEntity {
  property_id?: string;
  account_id: string;
  transaction_date: string;
  transaction_type: TransactionType;
  amount: string; // Decimal as string
  reference_number?: string;
  description?: string;
  memo?: string;
  tenant_id?: string;
  vendor_id?: string;
  invoice_id?: string;
}

// Budget interface
export interface Budget extends BaseEntity {
  property_id?: string;
  account_id: string;
  year: number;
  month: number;
  budgeted_amount: string; // Decimal as string
  notes?: string;
}

// Vendor interface
export interface Vendor extends BaseEntity {
  vendor_name: string;
  contact_person?: string;
  email?: string;
  phone?: string;
  address?: string;
  tax_id?: string;
  payment_terms?: string;
  is_active: boolean;
}

// Invoice interface
export interface Invoice extends BaseEntity {
  property_id?: string;
  vendor_id?: string;
  invoice_number: string;
  invoice_date: string;
  due_date: string;
  subtotal: string; // Decimal as string
  tax_amount: string; // Decimal as string
  total_amount: string; // Decimal as string
  amount_paid: string; // Decimal as string
  status: InvoiceStatus;
  notes?: string;
  line_items?: InvoiceLineItem[]; // Optional array of line items
}

// Invoice Line Item interface
export interface InvoiceLineItem {
  id: string;
  invoice_id: string;
  description: string;
  quantity: string; // Decimal as string
  unit_price: string; // Decimal as string
  amount: string; // Decimal as string
}

// Bank Account interface
export interface BankAccount extends BaseEntity {
  account_id: string;
  bank_name: string;
  account_number: string;
  routing_number?: string;
  account_type?: string;
  current_balance: string; // Decimal as string
  is_active: boolean;
}

// Extended interfaces with relationships
export interface InvoiceWithLineItems extends Invoice {
  line_items: InvoiceLineItem[];
}

export interface TransactionWithAccount extends Transaction {
  account?: Account;
}

export interface BudgetWithAccount extends Budget {
  account?: Account;
}

// API Response interfaces
export interface ApiResponse<T> {
  data: T;
}

export interface ApiListResponse<T> {
  data: T[];
}

// Request interfaces for creating/updating entities
export interface CreateAccountRequest {
  account_number: string;
  account_name: string;
  account_type: AccountType;
  parent_account_id?: string;
  description?: string;
  is_active?: boolean;
}

export interface UpdateAccountRequest {
  account_number?: string;
  account_name?: string;
  account_type?: AccountType;
  parent_account_id?: string;
  description?: string;
  is_active?: boolean;
}

export interface CreateTransactionRequest {
  property_id?: string;
  account_id: string;
  transaction_date: string;
  transaction_type: TransactionType;
  amount: string;
  reference_number?: string;
  description?: string;
  memo?: string;
  tenant_id?: string;
  vendor_id?: string;
  invoice_id?: string;
}

export interface UpdateTransactionRequest {
  property_id?: string;
  account_id?: string;
  transaction_date?: string;
  transaction_type?: TransactionType;
  amount?: string;
  reference_number?: string;
  description?: string;
  memo?: string;
  tenant_id?: string;
  vendor_id?: string;
  invoice_id?: string;
}

export interface CreateBudgetRequest {
  property_id?: string;
  account_id: string;
  year: number;
  month: number;
  budgeted_amount: string;
  notes?: string;
}

export interface UpdateBudgetRequest {
  property_id?: string;
  account_id?: string;
  year?: number;
  month?: number;
  budgeted_amount?: string;
  notes?: string;
}

export interface CreateVendorRequest {
  vendor_name: string;
  contact_person?: string;
  email?: string;
  phone?: string;
  address?: string;
  tax_id?: string;
  payment_terms?: string;
  is_active?: boolean;
}

export interface UpdateVendorRequest {
  vendor_name?: string;
  contact_person?: string;
  email?: string;
  phone?: string;
  address?: string;
  tax_id?: string;
  payment_terms?: string;
  is_active?: boolean;
}

export interface CreateInvoiceRequest {
  property_id?: string;
  vendor_id?: string;
  invoice_number: string;
  invoice_date: string;
  due_date: string;
  subtotal: string;
  tax_amount?: string;
  total_amount: string;
  amount_paid?: string;
  status?: InvoiceStatus;
  notes?: string;
  line_items: CreateInvoiceLineItemRequest[];
}

export interface UpdateInvoiceRequest {
  property_id?: string;
  vendor_id?: string;
  invoice_number?: string;
  invoice_date?: string;
  due_date?: string;
  subtotal?: string;
  tax_amount?: string;
  total_amount?: string;
  amount_paid?: string;
  status?: InvoiceStatus;
  notes?: string;
  line_items?: CreateInvoiceLineItemRequest[];
}

export interface CreateInvoiceLineItemRequest {
  description: string;
  quantity: string;
  unit_price: string;
  amount: string;
}

export interface CreateBankAccountRequest {
  account_id: string;
  bank_name: string;
  account_number: string;
  routing_number?: string;
  account_type?: string;
  current_balance?: string;
  is_active?: boolean;
}

export interface UpdateBankAccountRequest {
  account_id?: string;
  bank_name?: string;
  account_number?: string;
  routing_number?: string;
  account_type?: string;
  current_balance?: string;
  is_active?: boolean;
}

export interface UpdateBankAccountBalanceRequest {
  new_balance: string;
}

export interface MarkInvoicePaidRequest {
  payment_date: string;
  amount: string;
}

// Filter interfaces for API queries
export interface AccountFilters {
  account_type?: AccountType;
  is_active?: boolean;
}

export interface TransactionFilters {
  account_id?: string;
  property_id?: string;
  limit?: number;
  offset?: number;
}

export interface BudgetFilters {
  property_id?: string;
  year?: number;
  month?: number;
}

export interface VendorFilters {
  is_active?: boolean;
}

export interface InvoiceFilters {
  vendor_id?: string;
  property_id?: string;
  status?: InvoiceStatus;
}

export interface BankAccountFilters {
  is_active?: boolean;
}

// Report interfaces
export interface ProfitLossReport {
  period: string;
  property_id?: string;
  revenue: string;
  expenses: string;
  net_income: string;
  gross_profit_margin: string;
}

export interface BalanceSheetReport {
  as_of_date: string;
  property_id?: string;
  assets: string;
  liabilities: string;
  equity: string;
  total_liabilities_and_equity: string;
  balance_check: string;
}

export interface CashFlowReport {
  period: string;
  property_id?: string;
  operating_cash_flow: string;
  investing_cash_flow: string;
  financing_cash_flow: string;
  net_cash_flow: string;
}

export interface BudgetVsActualReport {
  period: string;
  property_id?: string;
  total_budgeted: string;
  total_actual: string;
  total_variance: string;
  budgets: BudgetVsActualItem[];
}

export interface BudgetVsActualItem {
  account_id: string;
  account_name: string;
  budgeted_amount: string;
  actual_amount: string;
  variance: string;
  variance_percentage: string;
}

export interface Vendor1099Data {
  vendor_id: string;
  vendor_name: string;
  tax_id?: string;
  year: number;
  total_invoices: number;
  total_amount: string;
  paid_amount: string;
  outstanding_amount: string;
  invoices: Vendor1099Invoice[];
}

export interface Vendor1099Invoice {
  invoice_id: string;
  invoice_number: string;
  invoice_date: string;
  total_amount: string;
  amount_paid: string;
  status: string;
}

// Report request interfaces
export interface ProfitLossRequest {
  property_id?: string;
  start_date: string;
  end_date: string;
}

export interface BalanceSheetRequest {
  property_id?: string;
  as_of_date: string;
}

export interface CashFlowRequest {
  property_id?: string;
  start_date: string;
  end_date: string;
}

export interface BudgetVsActualRequest {
  property_id?: string;
  year: number;
  month: number;
}

export interface Vendor1099Request {
  vendor_id: string;
  year: number;
}

File: backend/app/models/accounting.py
"""Accounting Models - Double-entry bookkeeping system"""

from sqlalchemy import Column, String, Integer, Numeric, DateTime, ForeignKey, Enum as SQLEnum, Text, Boolean
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from datetime import datetime
from uuid import uuid4
import enum
from . import Base

class AccountType(str, enum.Enum):
    ASSET = "asset"
    LIABILITY = "liability"
    EQUITY = "equity"
    REVENUE = "revenue"
    EXPENSE = "expense"

class TransactionType(str, enum.Enum):
    DEBIT = "debit"
    CREDIT = "credit"

class Account(Base):
    __tablename__ = "accounts"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    account_number = Column(String(50), unique=True, nullable=False, index=True)
    account_name = Column(String(255), nullable=False)
    account_type = Column(SQLEnum(AccountType), nullable=False)
    parent_account_id = Column(UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=True)
    description = Column(Text)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization", back_populates="accounts")
    parent = relationship("Account", remote_side=[id], backref="sub_accounts")
    transactions = relationship("Transaction", back_populates="account")

class Transaction(Base):
    __tablename__ = "transactions"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=True, index=True)
    account_id = Column(UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False, index=True)
    transaction_date = Column(DateTime, nullable=False, index=True)
    transaction_type = Column(SQLEnum(TransactionType), nullable=False)
    amount = Column(Numeric(12, 2), nullable=False)
    reference_number = Column(String(100), index=True)
    description = Column(Text)
    memo = Column(Text)
    tenant_id = Column(UUID(as_uuid=True), nullable=True)
    vendor_id = Column(UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=True)
    invoice_id = Column(UUID(as_uuid=True), ForeignKey("invoices.id"), nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization")
    property = relationship("Property")
    account = relationship("Account", back_populates="transactions")

class Budget(Base):
    __tablename__ = "budgets"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=True, index=True)
    account_id = Column(UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False, index=True)
    year = Column(Integer, nullable=False)
    month = Column(Integer, nullable=False)
    budgeted_amount = Column(Numeric(12, 2), nullable=False)
    notes = Column(Text)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization")
    property = relationship("Property")
    account = relationship("Account")

class Vendor(Base):
    __tablename__ = "vendors"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    vendor_name = Column(String(255), nullable=False)
    contact_person = Column(String(255))
    email = Column(String(255))
    phone = Column(String(50))
    address = Column(Text)
    tax_id = Column(String(50))
    payment_terms = Column(String(100))
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization")
    invoices = relationship("Invoice", back_populates="vendor")

class Invoice(Base):
    __tablename__ = "invoices"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    property_id = Column(UUID(as_uuid=True), ForeignKey("properties.id"), nullable=True, index=True)
    vendor_id = Column(UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=True)
    tenant_id = Column(UUID(as_uuid=True), nullable=True)
    invoice_number = Column(String(100), unique=True, nullable=False, index=True)
    invoice_date = Column(DateTime, nullable=False)
    due_date = Column(DateTime, nullable=False)
    subtotal = Column(Numeric(12, 2), nullable=False)
    tax_amount = Column(Numeric(12, 2), default=0)
    total_amount = Column(Numeric(12, 2), nullable=False)
    amount_paid = Column(Numeric(12, 2), default=0)
    status = Column(String(50), default="unpaid", index=True)
    notes = Column(Text)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization")
    property = relationship("Property")
    vendor = relationship("Vendor", back_populates="invoices")
    line_items = relationship("InvoiceLineItem", back_populates="invoice", cascade="all, delete-orphan")

class InvoiceLineItem(Base):
    __tablename__ = "invoice_line_items"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    invoice_id = Column(UUID(as_uuid=True), ForeignKey("invoices.id"), nullable=False)
    account_id = Column(UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False)
    description = Column(Text, nullable=False)
    quantity = Column(Numeric(10, 2), default=1)
    unit_price = Column(Numeric(12, 2), nullable=False)
    amount = Column(Numeric(12, 2), nullable=False)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    invoice = relationship("Invoice", back_populates="line_items")
    account = relationship("Account")

class BankAccount(Base):
    __tablename__ = "bank_accounts"
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    org_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    account_id = Column(UUID(as_uuid=True), ForeignKey("accounts.id"), nullable=False)
    bank_name = Column(String(255), nullable=False)
    account_number = Column(String(100), nullable=False)
    routing_number = Column(String(100))
    account_type = Column(String(50))
    current_balance = Column(Numeric(12, 2), default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, onupdate=datetime.utcnow)
    deleted_at = Column(DateTime)
    organization = relationship("Organization")
    account = relationship("Account")
EOF